docs/code/KeyValues.md

```diff
@@ -1,149 +1,155 @@
+  key values representing the characters observed in the string. The
+  `KeyValue` annotation "remembers" what each key is representing. (Note that
+  unlike many other key-valued operations, this uses a representation type of
+  `ushort` instead of `uint`.)
 
-* The `CharTokenizeTransform` will take input strings and produce key values
-  representing the characters observed in the string.
+* The `HashingTransformer` performs a hash of input values, and produces a key
+  value with count equal to the range of the hash function, which, if a `b`
+  bit hash was used, will produce values with a key-type of count `2ᵇ` .
+
+Note that in the first two cases, these are enumerating into a set, 
 
 ## Keys as Intermediate Values
 
 Explicitly invoking transforms that produce key values, and using those key
-values, is sometimes helpful. However, given that most trainers expect the
-feature vector to be a vector of floating point values and *not* keys, in
+values, is sometimes helpful. However, given that trainers typically expect
+the feature vector to be a vector of floating point values and *not* keys, in
 typical usage the majority of usages of keys is as some sort of intermediate
 value on the way to that final feature vector. (Unless, say, doing something
-like preparing labels for a multiclass learner.)
-
-So why not go directly to the feature vector, and forget this key stuff?
-Actually, to take text as the canonical example, we used to. However, by
-structuring the transforms from, say, text to key to vector, rather than text
-to vector *directly*, we are able to simplify a lot of code on the
-implementation side, which is both less for us to maintain, and also for users
-gives consistency in behavior.
-
-So for example, the `CharTokenize` above might appear to be a strange choice:
-*why* represent characters as keys? The reason is that the ngram transform is
-written to ingest keys, not text, and so we can use the same transform for
-both the n-gram featurization of words, as well as n-char grams.
+like preparing labels for a multiclass trainer.)
+
+So why not go directly to the feature vector from whatever the input was, and
+forget this key stuff? Actually, to take text processing as the canonical
+example, we used to. However, by structuring the transforms from, say, text to
+key to vector, rather than text to vector *directly*, we were able to make a
+more flexible pipeline, and re-use smaller, simpler components. Having
+multiple composable transformers instead of one "omni-bus" transformer that
+does everything makes the process easier to understand, maintain, and exploit
+for novel purposes, while giving people greater visibility into the
+composability of what actually happens.
+
+So for example, the `TokenizingByCharactersTransformer` above might appear to
+be a strange choice: *why* represent characters as keys? The reason is that
+the ngram transform which often comes after it  is written to ingest keys, not
```

Comment:
>which often comes after it  [](start = 20, length = 27)

nit: use comma before and after "which often comes after it"

Reply:
also nit: there are two spaces between `after it` and `is written`